### ISSUE TITLE
파라미터 값 최적화

### ISSUE THREAD

**Issue by uzinny04** [uzinny04/AI_Appl#3]
import os
import torch
from torchvision.datasets import ImageFolder
from torchvision.transforms import ToTensor
from PIL import Image
import numpy as np

# ------------------------------------------
# 1. 데이터셋의 특성을 자동으로 분석하는 함수
# ------------------------------------------
def analyze_dataset(path):
    dataset = ImageFolder(root=path, transform=ToTensor())
    num_classes = len(dataset.classes)
    num_samples = len(dataset)

    # 이미지 하나를 불러와 shape 분석
    sample_img, _ = dataset[0]
    c, h, w = sample_img.shape

    print("=== Dataset Analysis Result ===")
    print(f"- Total images      : {num_samples}")
    print(f"- Image size (CxHxW): {c} x {h} x {w}")
    print(f"- Number of classes : {num_classes}")

    return {
        "num_samples": num_samples,
        "num_classes": num_classes,
        "channels": c,
        "height": h,
        "width": w
    }

# ------------------------------------------
# 2. 분석 결과 기반으로 적정 파라미터 추천
# ------------------------------------------
def suggest_hyperparams(info):
    c = info["channels"]
    h = info["height"]
    w = info["width"]
    n = info["num_samples"]
    num_classes = info["num_classes"]

    # Batch size heuristic
    if h * w >= 200*200:
        batch = 4
    elif h * w >= 100*100:
        batch = 8
    else:
        batch = 16

    # Learning rate heuristic
    if n <= 200:
        lr = 1e-4
    elif n <= 1000:
        lr = 5e-5
    else:
        lr = 3e-5

    # Filter heuristic
    base_filter = 32 if h <= 128 else 64

    # Depth heuristic
    depth = 2 if h*w <= 64*64 else 3

    # Dropout heuristic
    dropout = 0.3 if n < 300 else 0.1

    print("\n=== Recommended Hyperparameters ===")
    print(f"Recommended batch_size    : {batch}")
    print(f"Recommended learning_rate : {lr}")
    print(f"Recommended base_filters  : {base_filter}")
    print(f"Recommended model_depth   : {depth} convolution blocks")
    print(f"Recommended dropout       : {dropout}")
    print(f"Classes (for fc2 output)  : {num_classes}")

    return {
        "batch_size": batch,
        "lr": lr,
        "base_filters": base_filter,
        "depth": depth,
        "dropout": dropout,
        "num_classes": num_classes
    }

# ------------------------------------------
# 3. 사용 예시 (경로만 바꿔서 실행)
# ------------------------------------------
train_path = "/content/drive/MyDrive/AI_Appl/train_data"
test_path  = "/content/drive/MyDrive/AI_Appl/test_data"

train_info = analyze_dataset(train_path)
params = suggest_hyperparams(train_info)

================================

✔ 출력 예시

예를 들어 데이터셋이 다음과 같다면:

3채널 RGB

64×128 이미지

클래스 2개

이미지 수 1200장

출력:

=== Dataset Analysis Result ===
- Total images      : 1200
- Image size (CxHxW): 3 x 64 x 128
- Number of classes : 2

=== Recommended Hyperparameters ===
Recommended batch_size    : 16
Recommended learning_rate : 3e-05
Recommended base_filters  : 32
Recommended model_depth   : 3 convolution blocks
Recommended dropout       : 0.1
Classes (for fc2 output)  : 2

이제 네 모델에서 바로 다음처럼 바꿀 수 있음:

batch_size = params["batch_size"]
learning_rate = params["lr"]
num_classes = params["num_classes"]
